osu.Game/Rulesets/Scoring/DrainingHealthProcessor.cs

```diff
@@ -60,23 +90,36 @@ protected override void Update()
         {
             base.Update();
 
-            if (!IsBreakTime.Value)
-            {
-                // When jumping in and out of gameplay time within a single frame, health should only be drained for the period within the gameplay time
-                double lastGameplayTime = Math.Clamp(Time.Current - Time.Elapsed, drainStartTime, gameplayEndTime);
-                double currentGameplayTime = Math.Clamp(Time.Current, drainStartTime, gameplayEndTime);
+            if (isInNonDrainSection)
+                return;
 
-                Health.Value -= drainRate * (currentGameplayTime - lastGameplayTime);
-            }
+            // When jumping in and out of gameplay time within a single frame, health should only be drained for the period within the gameplay time
+            double lastGameplayTime = Math.Clamp(Time.Current - Time.Elapsed, drainStartTime, gameplayEndTime);
+            double currentGameplayTime = Math.Clamp(Time.Current, drainStartTime, gameplayEndTime);
+
+            Health.Value -= drainRate * (currentGameplayTime - lastGameplayTime);
         }
 
         public override void ApplyBeatmap(IBeatmap beatmap)
         {
+            nonDrainSections.Clear();
+
             this.beatmap = beatmap;
 
             if (beatmap.HitObjects.Count > 0)
                 gameplayEndTime = beatmap.HitObjects[^1].GetEndTime();
 
+            // Ranges between the end of last hit object before a break
+            // and the start of first hit object after a break should
+            // not allow HP draining. (with break periods in)
+            foreach (BreakPeriod b in beatmap.Breaks)
+            {
+                var startTime = beatmap.HitObjects.LastOrDefault(h => h.GetEndTime() < b.StartTime)?.GetEndTime() ?? double.MinValue;
+                var endTime = beatmap.HitObjects.FirstOrDefault(h => h.StartTime > b.EndTime)?.StartTime ?? double.MaxValue;
```

Comment:
To avoid computing/accessing twice:
```suggestion
                var startTime = beatmap.HitObjects.Select(h => h.GetEndTime()).LastOrDefault(endTime => endTime < b.StartTime) ?? double.MinValue;
                var endTime = beatmap.HitObjects.Select(h => h.StartTime).FirstOrDefault(startTime => startTime > b.EndTime) ?? double.MaxValue;
```

Reply:
well, `StartTime` and `EndTime` are structures so default doesn't refer to `null` (`0.0d`), that's why I got forced into calculating twice...

Reply:
though tbh we may want to do something like a `FirstOrNull` & `LastOrNull` linq extensions (and even write it framework-side)

Reply:
Ehhh I forgot struct types do that. Could throw in a `Cast<double?>` but it's ugly too.

Helpers might not be necessary as this might get rectified when nullable reference types come in? Not sure.

Reply:
*actually it may get worse as now all normal declarations now will be considered non-nullables?*

Reply:
Actually it seems this can be written as:
```suggestion
                var startTime = beatmap.HitObjects.Select(h => h.GetEndTime()).Where(endTime => endTime < b.StartTime).DefaultIfEmpty(double.MinValue).Last();
                var endTime = beatmap.HitObjects.Select(h => h.StartTime).Where(startTime => startTime > b.EndTime).DefaultIfEmpty(double.MaxValue).First();
```
but that's a bit verbose too...